lnrpc/routerrpc/forward_interceptor.go

```diff
@@ -0,0 +1,259 @@
+	}
+	rpcInterceptor := interceptor
+	s.currentRPCInterceptor = rpcInterceptor
+	return nil
+}
+
+// onInterceptorDisconnected removes all previousely held forwards from
+// the store. Before they are removed it ensure to resume as the default
+// behavior.
+func (s *Server) onInterceptorDisconnected(interceptor *rpcInterceptor) {
+
+	log.Infof("RPC interceptor disconnected, resolving held packets")
+	interceptableForwarder := s.cfg.RouterBackend.InterceptableForwarder
+
+	// First de-register the current interceptor so we'll stop getting
+	// more interceptions.
+	interceptableForwarder.SetInterceptor(nil)
+
+	// Now drain and resume all previousely forwards.
+	interceptor.drain(func(forward htlcswitch.InterceptedForward) {
+		if err := forward.Resume(); err != nil {
+			log.Errorf("failed to resume hold forward %v", err)
+		}
+	})
+
+	// finally remove the current interceptor allows another one to connect.
+	s.rpcInterceptorMtx.Lock()
+	s.currentRPCInterceptor = nil
+	s.rpcInterceptorMtx.Unlock()
+}
+
+// rpcInterceptor is a helper struct that handles the lifecycle of an rpc
+// interceptor streaming session.
+// It is created when the stream opens and drains when the stream closes.
+type rpcInterceptor struct {
+	sync.Mutex
+
+	// holdForwards is a map of current hold forwards and their corresponding
+	// ForwardResolver.
+	holdForwards map[channeldb.CircuitKey]htlcswitch.InterceptedForward
+
+	// stream is the bidirectional stream
+	stream Router_HtlcInterceptorServer
+}
+
+// newRPCInterceptor creates a new rpcInterceptor.
+func newRPCInterceptor(stream Router_HtlcInterceptorServer) *rpcInterceptor {
+	return &rpcInterceptor{
+		stream: stream,
+		holdForwards: make(
+			map[channeldb.CircuitKey]htlcswitch.InterceptedForward),
+	}
+}
+
+// hold just store the intercepted forward.
+func (r *rpcInterceptor) hold(forward htlcswitch.InterceptedForward) {
+	r.Lock()
+	r.holdForwards[forward.CircuitKey()] = forward
+	r.Unlock()
+}
+
+// forwardToClient forwards the intercepted htlc to the client.
+func (r *rpcInterceptor) forwardToClient(forward htlcswitch.InterceptedForward) error {
+	htlc := forward.Packet()
+	inKey := forward.CircuitKey()
+
+	// A ChannelAcceptRequest has been received, send it to the client.
+	interceptionRequest := &ForwardHtlcInterceptRequest{
+		CircuitKey: &CircuitKey{
+			IncomingChanId: inKey.ChanID.ToUint64(),
+			HtlcId:         inKey.HtlcID,
+		},
+		HtlcPaymentHash: htlc.PaymentHash[:],
+		AmountSat:       uint64(htlc.Amount.ToSatoshis()),
+		Expiry:          htlc.Expiry,
+	}
+
+	return r.stream.Send(interceptionRequest)
+}
+
+// resolveFromClient handles a resolution arrived from the client.
+func (r *rpcInterceptor) resolveFromClient(in *ForwardHtlcInterceptResponse) error {
+	circuitKey := channeldb.CircuitKey{
+		ChanID: lnwire.NewShortChanIDFromInt(in.CircuitKey.IncomingChanId),
+		HtlcID: in.CircuitKey.HtlcId,
+	}
+	var interceptedForward htlcswitch.InterceptedForward
+	r.Lock()
```

Comment:
Is there a way to avoid this lock? There is already a main message loop to synchronize access

Reply:
Yes, good one. I moved the previous `hold` call to the main loop and now the `rpcInterceptor` is only used within the main loop. I also added a comment that it should only be called from one go routine.